dice_ml/explainer_interfaces/dice_KD.py

```diff
@@ -0,0 +1,236 @@
+"""
+Module to generate diverse counterfactual explanations based on tensorflow 2.x
+"""
+from dice_ml.explainer_interfaces.explainer_base import ExplainerBase
+import tensorflow as tf
+
+import numpy as np
+import random
+import timeit
+from sklearn.neighbors import KDTree
+import pandas as pd
+
+from dice_ml import diverse_counterfactuals as exp
+
+class DiceKD(ExplainerBase):
+
+    def __init__(self, data_interface, model_interface):
+        """Init method
+
+        :param data_interface: an interface class to access data related params.
+        :param model_interface: an interface class to access trained ML model.
+
+        """
+
+        super().__init__(data_interface) # initiating data related parameters
+
+        # initializing model variables
+        self.model = model_interface
+
+        # loading trained model
+        self.model.load_model()
+
+        # variables required to generate CFs - see generate_counterfactuals() for more info
+        self.cfs = []
+        self.features_to_vary = []
+        self.cf_init_weights = []  # total_CFs, algorithm, features_to_vary
+        self.loss_weights = []  # yloss_type, diversity_loss_type, feature_weights
+        self.feature_weights_input = ''
+        self.hyperparameters = [1, 1, 1]  # proximity_weight, diversity_weight, categorical_penalty
+        self.optimizer_weights = []  # optimizer, learning_rate
+
+        # number of output nodes of ML model
+        temp_input = tf.convert_to_tensor([tf.random.uniform([len(self.data_interface.encoded_feature_names)])], dtype=tf.float32)
```

Comment:
can this line be written without explicitly invoking tensorflow? One option is to implement a function called `provide_temp_input` inside the model class `self.model.`

That way, this same code can work for all types of models and you don't need to write a different explainer for tensorflow, sklearn etc. The model interface will implement any model specific method.

Reply:
in fact, both lines can be folded (43 and 44) into a single function `get_num_output_nodes()` inside `self.model`, which does exactly these two operations.